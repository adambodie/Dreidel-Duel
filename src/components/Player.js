import React, { Component } from "react";
import styles from "./player.module.css";


export default class Player extends Component {
  constructor(props) {
    super(props);
    this.state = {
		value: ''
	}
      
	this.handleChange = this.handleChange.bind(this);
	this.handleNameChange = this.handleNameChange.bind(this);
  }
	handleChange(e) {
		this.props.onScoreChange(e.target.value);  
  }
	handleNameChange(e) {
		this.setState({value: e.target.value});
	}

  render() {
    return (
		<div className={styles.player}>
			<h2>Player {this.props.id}</h2>
			<input type="text" value={this.state.value} onChange={this.handleNameChange} />
			<div className={styles.counter}>
				<button
					className={styles.counterAction}
					onClick={this.handleChange}
					disabled={this.props.disabled}
					>
					SPIN
				</button>
				<div className={styles.counterScore}>
					{this.props.score}
				</div>
			</div>
		</div>
				
    );
  }
}
