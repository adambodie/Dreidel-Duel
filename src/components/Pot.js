import React from "react"

const Pot = ({pot}) => {
	return(
		<div>
			<h3 style={{fontFamily: 'Vollkorn', margin: 0}}>Coins in the Pot:</h3>
			<h2 style={{fontFamily: 'Vollkorn', fontSize: '2em'}}>{pot}</h2>
		</div>
	)
}

export default Pot;
