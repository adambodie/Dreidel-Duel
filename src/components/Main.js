// Libs
import React, { Component } from "react";
import styles from "../css/main.module.css";
import Pot from './Pot';
import Spin from './Spin';
import Reset from './Reset';
import Start from './Start';
import Add from './Add';
import Player from './Player';


export default class Main extends Component {
	constructor() {
		super();
		this.state = {
			pot: 10,
			spin: "",
			turn: 0,
			isDisabled: true,
			isButtonDisabled: true,
			players: [
				{ score: 5, isDisabled: true, id: 1 },
				{ score: 5, isDisabled: true, id: 2 }
			]
		};
		this.onStartChange = this.onStartChange.bind(this);
		this.onResetChange = this.onResetChange.bind(this);
		this.onScoreChange = this.onScoreChange.bind(this);
		this.onAddChange = this.onAddChange.bind(this);
	}

	onStartChange() {
		const newPlayers = [...this.state.players];
		newPlayers[0].isDisabled = false;
		this.setState(prevState => ({
			newPlayers,
			isDisabled: !prevState.isDisabled
		}));
	}
	
	onResetChange() {
		const newPlayers = [...this.state.players];
		newPlayers.forEach((item) => {
			item.isDisabled = true;
			item.score = 5;
		});
		this.setState(prevState => ({
			newPlayers,
			isDisabled: !prevState.isDisabled,
			pot: 10,
			spin: "",
			isButtonDisabled: true,
			turn: 0
		}));
	}
  
	onAddChange() {
		const newPlayers = [...this.state.players];
		let playerOne = newPlayers[0];
		let playerTwo = newPlayers[1];
		if (playerOne.score === 0 || playerTwo.score === 0) {
			newPlayers.forEach((element) => element.isDisabled = true );
			this.setState(prevState => ({
				newPlayers,
				pot: "Game Over, play again"
			}));
		} else {
			if (playerOne.isDisabled === true &&  this.state.turn >= 1) {
				newPlayers.forEach((element) => element.score -= 1 );
				playerOne.isDisabled = false;
				this.setState(prevState => ({ pot: this.state.pot + 2 }));
			}
		}
		this.setState(prevState => ({ isButtonDisabled: true }));
	}
  
	onScoreChange() {
		let dreidel = Math.floor(Math.random() * 4);
		const hebrew = ["נ", "שׁ", "ה", "ג"];
		let currentSpin = hebrew[dreidel];
		const newPlayers = [...this.state.players];
		let playerOne = newPlayers[0];
		let playerTwo = newPlayers[1];

		if (playerOne.isDisabled === true && this.state.turn >= 1) {
			this.setState(prevState => ({ isButtonDisabled: false }));
		}
		if ((playerOne.score <= 0 || playerTwo.score <= 0) && (dreidel === 1 || dreidel === 3)) {
			newPlayers.forEach((element) => element.isDisabled = true );
			this.setState(prevState => ({
				newPlayers,
				pot: "Game Over, play again"
			}));
		} else {
			if (dreidel === 0) {
			/* Spin a Nun - Nothing */
				if (playerOne.isDisabled === false) {
					newPlayers.forEach((element) => element.isDisabled = !element.isDisabled );
				} else {
					playerTwo.isDisabled = true;
				}
			} else if (dreidel === 1) {
			/* Spin a Shin - Put one in the Pot */
				if (playerOne.isDisabled === false) {
					newPlayers.forEach((element) => element.isDisabled = !element.isDisabled );
					playerOne.score -= 1;
				} else {
					playerTwo.score -= 1;
					playerTwo.isDisabled = true;
				}
				this.setState(prevState => ({ pot: this.state.pot + 1 }));
			} else if (dreidel === 2) {
			/* Spin a Hay - Win half of your coins in the Pot */
				if (playerOne.isDisabled === false) {
					newPlayers.forEach((element) => element.isDisabled = !element.isDisabled );
					playerOne.score += Math.round(this.state.pot / 2);
				} else {
					playerTwo.score += Math.round(this.state.pot / 2);
					playerTwo.isDisabled = true;
				}
				this.setState(prevState => ({ pot: Math.floor(this.state.pot / 2) }));
			} else {
			/* Spin a Gimel - Win everything  */
				if (playerOne.isDisabled === false) {
					playerOne.score += this.state.pot;
					newPlayers.forEach((element) => element.isDisabled = !element.isDisabled );
				} else {
					playerTwo.score += this.state.pot;
					playerTwo.isDisabled = true;
				}
				newPlayers.forEach((element) => element.score  -= 1 );
				this.setState(prevState => ({ pot: 2 }));
			}
			this.setState(prevState => ({
				newPlayers,
				spin: currentSpin,
				turn: this.state.turn + 1
			}));
		}
	}

  render() {
    return (
		<div className={styles.container}>
			<div className={styles.players}>
				<Player 
					score={this.state.players[0].score} 
					disabled={this.state.players[0].isDisabled} 
					onScoreChange={this.onScoreChange} 
					players={this.state.players}
					id={this.state.players[0].id}
					/>
					<Spin spin={this.state.spin} />
				<Player  
					score={this.state.players[1].score} 
					disabled={this.state.players[1].isDisabled} 
					onScoreChange={this.onScoreChange} 
					players={this.state.players}
					id={this.state.players[1].id}
					/>
			</div>
			<div className={styles.footer}>
				<Add onAddChange={this.onAddChange} disabled={this.state.isButtonDisabled}/>
				<Pot pot={this.state.pot} />
			</div>
			<div className={styles.buttons}>
				<Reset onResetChange={this.onResetChange} disabled={this.state.isDisabled} />
				<Start onStartChange={this.onStartChange} disabled={!this.state.isDisabled} />
			</div>
        </div>
    );
  }
}
