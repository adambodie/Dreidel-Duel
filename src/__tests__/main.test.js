import React from "react"
import renderer from "react-test-renderer"
import Main from "../components/Main"
import Player from "../components/Player"
import Pot from "../components/Pot"
import Start from "../components/Start"
import {shallow} from "enzyme"

describe("Main", () =>
	it("renders correctly", () => {
		const tree = renderer.create(<Main />).toJSON()
		expect(tree).toMatchSnapshot()
	}),
	it('renders two <Player /> components', () => {
		const wrapper = shallow(<Main />);
		expect(wrapper.find(Player).length).toBe(2);
	}),
	it('has a isDisabled state of true', () => {
		const wrapper = shallow(<Main />);
		expect(wrapper.state().isDisabled).toEqual(true);
	})
	
)
