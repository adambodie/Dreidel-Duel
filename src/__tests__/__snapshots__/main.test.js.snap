// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="players"
  >
    <div
      className="player"
    >
      <h2>
        Player 
        1
      </h2>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <div
        className="counter"
      >
        <button
          className="counterAction"
          disabled={true}
          onClick={[Function]}
        >
          SPIN
        </button>
        <div
          className="counterScore"
        >
          5
        </div>
      </div>
    </div>
    <h2
      style={
        Object {
          "fontSize": 64,
          "width": 55,
        }
      }
    >
      
    </h2>
    <div
      className="player"
    >
      <h2>
        Player 
        2
      </h2>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <div
        className="counter"
      >
        <button
          className="counterAction"
          disabled={true}
          onClick={[Function]}
        >
          SPIN
        </button>
        <div
          className="counterScore"
        >
          5
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <button
      className="action"
      disabled={true}
      onClick={[Function]}
    >
      Add Coin
    </button>
    <div>
      <h3>
        Coins in the Pot:
      </h3>
      <h2>
        10
      </h2>
    </div>
  </div>
  <div
    className="buttons"
  >
    <button
      className="action"
      disabled={true}
      onClick={[Function]}
    >
      Reset
    </button>
    <button
      className="action"
      disabled={false}
      onClick={[Function]}
    >
      Start
    </button>
  </div>
</div>
`;
